e2e/nomostest/gitproviders/repository.go

```diff
@@ -248,82 +279,80 @@ func (g *Repository) init(privateKey string) {
 //
 // Don't put multiple manifests in the same file unless parsing multi-manifest
 // files is the behavior under test. In that case, use AddFile.
-func (g *Repository) Add(path string, obj client.Object) {
-	g.T.Helper()
+func (g *Repository) Add(path string, obj client.Object) error {
 	testkubeclient.AddTestLabel(obj)
-	// TODO: Figure out how to cleanly inject runtime.Scheme here.
 
-	// We have to make a pass through json since yaml.Marshal does not respect
-	// json "omitempty" directives.
-	var bytes []byte
-	var err error
-	var u *unstructured.Unstructured
 	ext := filepath.Ext(path)
 	switch ext {
 	case ".yaml", ".yml":
 		// We must convert through JSON/Unstructured to avoid "omitempty" fields
 		// from being specified.
-		u, err = reconcile.AsUnstructuredSanitized(obj)
+		var err error
+		uObj, err := reconcile.AsUnstructuredSanitized(obj)
+		if err != nil {
+			return errors.Wrap(err, "sanitizing object")
+		}
+		// Encode converts to json then yaml, to handle "omitempty" directives.
+		bytes, err := runtime.Encode(g.yamlSerializer, uObj)
 		if err != nil {
-			g.T.Fatal(err)
+			return errors.Wrap(err, "encoding object as yaml")
 		}
-		bytes, err = yaml.Marshal(u)
+		return g.AddFile(path, bytes)
 	case ".json":
-		u, err = reconcile.AsUnstructuredSanitized(obj)
+		var err error
+		uObj, err := reconcile.AsUnstructuredSanitized(obj)
 		if err != nil {
-			g.T.Fatal(err)
+			return errors.Wrap(err, "sanitizing object")
 		}
-		bytes, err = json.MarshalIndent(u, "", "  ")
+		bytes, err := json.MarshalIndent(uObj, "", "  ")
+		if err != nil {
+			return errors.Wrap(err, "encoding object as json")
+		}
+		return g.AddFile(path, bytes)
 	default:
 		// If you're seeing this error, use "AddFile" instead to test ignoring
 		// files with extensions we ignore.
-		err = fmt.Errorf("invalid extension to write object to, %q, use .AddFile() instead", ext)
-	}
-	if err != nil {
-		g.T.Fatal(err)
+		return fmt.Errorf("invalid extension to write object to, %q, use .AddFile() instead", ext)
 	}
-
-	g.AddFile(path, bytes)
```

Comment:
How come this shared call was moved up to each of the cases? This seemed fine with the outer scoped `bytes` declaration

Reply:
Just a matter of taste. When I replaced the Fatal calls with returns I realized that some of the switch cases were returning inside the switch and some weren't. So I move all the returns into the switch.

Yes, it means this AddFile call is duplicated in 2 cases, but it has some benefits:
1. The compiler now validates that all possible cases return
2. It's more obvious that the default case cannot ever lead to the AddFile code branch.
3. It becomes more clear that AsUnstructuredSanitized is not returning just an error, otherwise we wouldn't need the `var err error` to down-cast it.
4. We don't need the `var bytes []byte` or `var u *unstructured.Unstructured` outside the switch.

I could change it back, if you prefer, but I think this is slightly better.